stdlib/universe/double_exponential_moving_average.go

```diff
@@ -0,0 +1,406 @@
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (t *DEMATransformation) UpdateWatermark(id execute.DatasetID, mark execute.Time) error {
+	return t.d.UpdateWatermark(mark)
+}
+
+func (t *DEMATransformation) UpdateProcessingTime(id execute.DatasetID, pt execute.Time) error {
+	return t.d.UpdateProcessingTime(pt)
+}
+
+func (t *DEMATransformation) Finish(id execute.DatasetID, err error) {
+	t.d.Finish(err)
+}
+
+func (t *DEMATransformation) passThrough(vs *moving_average.ArrayContainer, b execute.TableBuilder, bj int) error {
+	// This is faster than passing through EMA twice
+	j := 0
+
+	for ; int64(t.i[bj]) < 2*t.n-1 && j < vs.Len(); t.i[bj]++ {
+		if vs.IsNull(j) {
+			t.lastVal[bj] = nil
+		} else {
+			t.lastVal[bj] = vs.OrigValue(j)
+		}
+		if int64(t.i[bj]) == 2*t.n-2 {
+			if vs.IsNull(j) {
+				if err := b.AppendNil(bj); err != nil {
+					return err
+				}
+			} else {
+				if err := b.AppendValue(bj, values.New(vs.OrigValue(j))); err != nil {
+					return err
+				}
+			}
+		}
+		j++
+	}
+
+	for ; int64(t.i[bj]) >= 2*t.n-1 && j < vs.Len(); t.i[bj]++ {
+		if vs.IsNull(j) {
+			if err := b.AppendNil(bj); err != nil {
+				return err
+			}
+		} else {
+			if err := b.AppendValue(bj, values.New(vs.OrigValue(j))); err != nil {
+				return err
+			}
+		}
+		j++
+	}
+	return nil
+}
+
+func (t *DEMATransformation) doFirstEMA(vs *moving_average.ArrayContainer, b execute.TableBuilder, bj int, doDEMA bool) error {
+	// if !doDEMA, append the last 2n - 1 values
+	if !doDEMA {
+		return t.passThrough(vs, b, bj)
+	}
+
+	return t.ema1.DoNumeric(vs, b, bj, doDEMA, false)
+}
+
+func (t *DEMATransformation) passThroughTime(vs *array.Int64, b execute.TableBuilder, bj int) error {
+	j := 0
+
+	for ; int64(t.i[bj]) < 2*t.n-1 && j < vs.Len(); t.i[bj]++ {
+		if vs.IsNull(j) {
+			t.lastVal[bj] = nil
+		} else {
+			t.lastVal[bj] = execute.Time(vs.Value(j))
+		}
+		if int64(t.i[bj]) == 2*t.n-2 {
+			if vs.IsNull(j) {
+				if err := b.AppendNil(bj); err != nil {
+					return err
+				}
+			} else {
+				if err := b.AppendTime(bj, execute.Time(vs.Value(j))); err != nil {
+					return err
+				}
+			}
+		}
+
+		j++
+	}
+
+	for ; int64(t.i[bj]) >= 2*t.n-1 && j < vs.Len(); t.i[bj]++ {
+		if vs.IsNull(j) {
+			if err := b.AppendNil(bj); err != nil {
+				return err
+			}
+		} else {
+			if err := b.AppendTime(bj, execute.Time(vs.Value(j))); err != nil {
+				return err
+			}
+		}
+		j++
+	}
+	return nil
+}
+
+func (t *DEMATransformation) doSecondEMA(b execute.TableBuilder, bj int) error {
+	// Get the first EMA
+	firstEMA := t.ema1.GetEMA(bj)
+
+	// Convert firstEMA to *array.Float64
+	vals := make([]float64, len(firstEMA))
+	valid := make([]bool, len(firstEMA))
+	for i, val := range firstEMA {
+		if val != nil {
+			vals[i] = val.(float64)
+			valid[i] = true
+		} else {
+			vals[i] = 0
+			valid[i] = false
+		}
+	}
+
+	bld := arrow.NewFloatBuilder(t.alloc)
+	defer bld.Release()
+
+	bld.AppendValues(vals, valid)
```

Comment:
did you consider building up the arrow array incrementally instead of filling up a slice then calling this?

Reply:
I looked at Jonathan's arrow builder, but as it is right now it doesn't let you append nulls.

Reply:
I'll fix that up and recommit